perezjln/gym-lowcostrobot
# Added new teleoperation script + PushCubeLoop env + small refactoring of the environments.

## What does this PR do:
- Added new teleoperation script with LeRobot code in `examples/teleoperate_leader_lerobot.py`
- Added new `PushCubeLoop-v0` env in `push_cube_loop_env.py` to mimic the RL task currently used in LeRobot.
- Added visualizations  to `PickPlace-v0` and `PushCube-v0` in order to visualize the target region. 
- Added indexing to the cube and arm poses based on their body name in the sim in all envs rather than constant indices.
- Added control decimation, i.e. multiple simulation steps per control step.
- Improved the contact between gripper and cube by modifying the collision params in the description. 

## Bug fixes:
- Fixed bug in teleoperation by sending control through mujoco rather than modifying joint in data.
- Fixed index issue as after #33 the order of the robot and cubes was inverted so the indices of the joint and cube were completely wrong. 
- Fixed robot orientation. The new robot description added in #33 was rotated by 90 degrees so the robot was facing the cube in all environments.

## How to checkout and test:
In order to test the teleoperation you should:
1. Get the `start_pos` from the calibration done by LeRobot and change the values in `examples/teleoperate_leader_lerobot.py`. (can be found in `.cache/calibration/koch/main_leader.json`)
2. Run (use mjpython for mac):
```
python examples/teleoperate_leader_lerobot.py --device /dev/tty.usbmodem585A0078211 --env-name PushCubeLoop-v0
```
depending on your device and environment you'd like to test.

## TODO
### Learning:
- [ ] Validate action space: instead of absolute joint angle commands add a delta joint angle commands.
- [ ] Validate reward function: debug reward function by trying to learn the tasks with LeRobot's tdmpc
- [ ] Add sparse reward functions.
